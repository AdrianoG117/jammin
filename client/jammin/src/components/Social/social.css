.social-container {
  display: flex;
  flex-direction: column;
  /* align-items: flex-end;
  justify-content: center; */
  height: 65%;
}

.msg-container {
  /* background-color: red; */
  height: 80%;
  width: 100%;
  overflow-y: scroll;
  padding: 1rem;
}

.msg-indiv {
  background-color: rgba(255, 255, 255, 0.9);
  color: black;
  margin: 1rem;
  padding: 0.3rem;
  border-radius: 13px;
}

.msg-name {
  font-size: 1rem;
  /* background-color: red; */
}

.msg-message {
  /* background-color: yellow; */
  font-size: 1.5rem;
}

.form-container {
  /* background-color: yellow; */
  height: 20%;
  width: 100%;
}

.social-form {
  display: flex;
  width: 100%;
  height: 100%;
}

.left-form {
  display: flex;
  flex-direction: column;
  width: 80%;
  height: 100%;
}

.right-form {
  width: 20%;
  display: flex;
  align-items: center;
  justify-content: center;
}

.social-input {
  background-color: rgba(0, 0, 0, 0.7);
  color: white;
  padding: 0.4rem;
}

.social-input::placeholder {
  color: white;
}

#name-input {
  height: 25%;
}

#message-input {
  height: 75%;
}

#post-msg {
  height: 100%;
  width: 100%;
  background-color: rgba(0, 0, 0, 0.7);
  color: white;
}

#post-msg:hover {
  background-color: rgba(255, 255, 255, 1);
  color: black;
}
