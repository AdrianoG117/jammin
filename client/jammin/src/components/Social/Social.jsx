import React, { useEffect, useRef } from 'react';
import './social.css';
import apiService from '../../ApiService';

function Social({ jam, msg, setMsg, initialState, isSignedUp, userData }) {
  const messagesEndRef = useRef(null);
  const dummyDiv = useRef();

  const scrollToBottom = () => {
    dummyDiv.current?.scrollIntoView({ behavior: 'smooth' });
  };

  useEffect(() => {
    if (messagesEndRef) {
      messagesEndRef.current.addEventListener('DOMNodeInserted', (event) => {
        const { currentTarget: target } = event;
        target.scroll({ top: target.scrollHeight, behavior: 'smooth' });
      });
    }
    scrollToBottom();
  }, []);

  function handleChange(e) {
    setMsg((previous) => ({
      name: userData.firstname,
      message: e.target.value,
    }));
  }

  async function handleSubmit(e) {
    e.preventDefault();
    await apiService.postMessage(msg, jam._id);
    setMsg(initialState);
  }

  return (
    <div className="social-container">
      <div className="msg-container" ref={messagesEndRef}>
        {jam.messages.map((msg) => (
          <div className="msg-indiv">
            <p className="msg-name">{msg.name}</p>
            <p className="msg-message">{msg.message}</p>
          </div>
        ))}
        <div ref={dummyDiv}></div>
      </div>
      <div className="form-container">
        {isSignedUp ? (
          <form className="social-form" onSubmit={handleSubmit}>
            <div className="left-form">
              <textarea
                required
                type="text"
                placeholder="MESSAGE"
                name="message"
                value={msg.message}
                onChange={handleChange}
                className="social-input"
                id="message-input"
              ></textarea>
            </div>
            <div className="right-form">
              <button id="post-msg">POST</button>
            </div>
          </form>
        ) : (
          <div className="social-err">
            <h2>Log in to use the chat !</h2>
          </div>
        )}
      </div>
    </div>
  );
}

export default Social;
