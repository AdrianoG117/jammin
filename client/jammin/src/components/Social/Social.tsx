import React, { useEffect, useRef, FunctionComponent} from "react";
import "./social.css";
import apiService from "../../apiService/ApiService";
import {Message, Jam, User} from "../../apiService/APIResponseTypes";

interface IProps {
  userData: User 
  setUserData: React.Dispatch<React.SetStateAction< User>>
  isSignedUp: boolean
  jam: Jam
  msg: Message
  setMsg: React.Dispatch<React.SetStateAction<Message>>
  initialState: Message

}

const  Social:FunctionComponent<IProps> = ({ jam, msg, setMsg, initialState, isSignedUp, userData }:IProps) => {
  const messagesEndRef = useRef<HTMLDivElement>(null);
  const dummyDiv = useRef<HTMLDivElement>(null);

  const scrollToBottom = () => {
    dummyDiv.current?.scrollIntoView({ behavior: "smooth" });
  };

  useEffect(() => {
    if (messagesEndRef) {
      messagesEndRef.current?.addEventListener("DOMNodeInserted", (event) => {
        const { currentTarget: target } = event ;
        if (target){
          (target as HTMLDivElement).scroll({ top: (target as HTMLDivElement).scrollHeight, behavior: "smooth" });
        }
      });
    }
    scrollToBottom();
  }, []);

  function handleChange(e:React.ChangeEvent<HTMLTextAreaElement>): void {
    setMsg((previous) => ({
      ...previous,
      name: userData.firstname,
      message: e.currentTarget.value,

    }));
  }

  async function handleSubmit(e:React.FormEvent<HTMLFormElement>):Promise<void> {
    e.preventDefault();
    if(jam._id) await apiService.postMessage(msg, jam._id);
    setMsg(initialState);
  }

  return (
    <div className="social-container">
      <div className="msg-container" ref={messagesEndRef}>
        {jam.messages.map((msg) => (
          <div className="msg-indiv">
            <p className="msg-name">{msg.name}</p>
            <p className="msg-message">{msg.message}</p>
          </div>
        ))}
        <div ref={dummyDiv}></div>
      </div>
      <div className="form-container">
        {isSignedUp ? (
          <form className="social-form" onSubmit={handleSubmit}>
            <div className="left-form">
              <textarea
                required            
                placeholder="MESSAGE"
                name="message"
                value={msg.message}
                onChange={handleChange}
                className="social-input"
                id="message-input"
              ></textarea>
            </div>
            <div className="right-form">
              <button id="post-msg">POST</button>
            </div>
          </form>
        ) : (
          <div className="social-err">
            <h2>Log in to use the chat !</h2>
          </div>
        )}
      </div>
    </div>
  );
};

export default Social;
