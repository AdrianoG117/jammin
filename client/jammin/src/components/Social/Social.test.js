import {render, screen} from "@testing-library/react";
import Social from "./Social";


describe("Social component", ()=> {

   const scrollIntoView = window.HTMLElement.prototype.scrollIntoView = jest.fn()
  
  test("should match the snapshot", () => {
    const {container} = render(<Social
                  userData={{}}
                  setUserData={()=>{}}
                  isSignedUp={false}
                  setIsSignedUp={()=>{}}
                  jam={{}}
                  msg={{name:"", message: ""}}
                  setMsg={()=>{}}
                  initialState={{}}
                  scrollIntoView={scrollIntoView}
                   />);
    expect(container.firstChild).toMatchSnapshot();
  });
  
  test("Should render the chat box and post button if signed in", () => {
    render (<Social
                  userData={{}}
                  setUserData={()=>{}}
                  isSignedUp={true}
                  setIsSignedUp={()=>{}}
                  jam={{}}
                  msg={{name:"", message: ""}}
                  setMsg={()=>{}}
                  initialState={{}} 
                  scrollIntoView={scrollIntoView}
                  />);
    //title
    screen.getByPlaceholderText(/MESSAGE/);
    screen.getByRole("button", { name: "POST" });
  });

  test("Should message user to sign in if not signed in", () => {
    render (<Social
                  userData={{}}
                  setUserData={()=>{}}
                  isSignedUp={false}
                  setIsSignedUp={()=>{}}
                  jam={{}}
                  msg={{name:"", message: ""}}
                  setMsg={()=>{}}
                  initialState={{}} 
                  scrollIntoView={scrollIntoView}
                  />);
    //title
    screen.getByText(/Log in to use the chat !/);
  });

});