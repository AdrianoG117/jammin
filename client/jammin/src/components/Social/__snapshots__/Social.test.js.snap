// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Social component should match the snapshot 1`] = `
<div
  class="social-container"
>
  <div
    class="msg-container"
  >
    <div />
  </div>
  <div
    class="form-container"
  >
    <div
      class="social-err"
    >
      <h2>
        Log in to use the chat !
      </h2>
    </div>
  </div>
</div>
`;
